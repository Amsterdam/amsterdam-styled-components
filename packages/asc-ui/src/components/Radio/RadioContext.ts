import React from 'react'

const RadioContext = React.createContext({
  checked: '',
  setChecked: () => {},
  name: undefined,
  error: false,
} as
  {
    checked: string
    setChecked: Function
    name: string | undefined
    error: boolean | undefined
  })
export default RadioContext
