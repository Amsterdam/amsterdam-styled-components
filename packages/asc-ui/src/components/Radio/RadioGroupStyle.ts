import styled, { css } from '@datapunt/asc-core'
import LabelStyle from '../Label/LabelStyle'
import { themeSpacing } from '../../utils'

export type Props = {
  name?: string
  horizontal?: boolean
  disabled?: boolean
}

export default styled.div<Props>`
  display: flex;
  flex-direction: ${({ horizontal }) => (horizontal ? 'row' : 'column')};
  ${({ horizontal }) =>
    horizontal &&
    css`
      ${LabelStyle} {
        margin-right: ${themeSpacing(4)};
      }
    `}
`
