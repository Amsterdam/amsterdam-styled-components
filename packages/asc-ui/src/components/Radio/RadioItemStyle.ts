import styled from '@datapunt/asc-core'

const RadioItemStyle = styled.div`
  position: relative;
`
export default RadioItemStyle
