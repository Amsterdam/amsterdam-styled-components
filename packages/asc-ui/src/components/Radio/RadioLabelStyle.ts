import styled from '@datapunt/asc-core'

interface Props {
  htmlFor: string
  disabled?: boolean
}

const RasdioLabelStyle = styled.label.attrs<Props>({
  htmlFor: ({ htmlFor }: Props) => htmlFor,
  disabled: ({ disabled }: Props) => disabled,
})<Props>`
  color: ${props => (props.disabled ? '#bebebe' : '#434343')};
  font-family: 'AvenirNextLTW01-Regular', verdana, sans-serif;
  font-size: 16px;
  padding-left: 38px;
  line-height: 28px;

  &::before {
    border-radius: 50%;
    border: ${props =>
      props.disabled ? '1px solid #bebebe' : '1px solid #000'};
    content: '';
    height: calc(28px - 4px);
    left: 0;
    position: absolute;
    top: 0;
    width: calc(28px - 4px);
    zoom: 1;
  }

  &::after {
    background-color: #0000;
    border: 6px solid;
    border-radius: 50%;
    color: ${props => (props.disabled ? '#bebebe' : '#000')};
    content: '';
    height: 0;
    left: 7px;
    opacity: 0;
    position: absolute;
    top: 7px;
    width: 0;
    zoom: 1;
  }

  input[type='radio']:checked + &::after {
    opacity: 1;
    zoom: 1;
  }
`
export default RasdioLabelStyle
