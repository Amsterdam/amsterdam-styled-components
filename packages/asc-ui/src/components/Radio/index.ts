export { default as RadioStyle, RadioWrapperStyle } from './RadioStyle'
export { default } from './Radio'
export { default as RadioGroup } from './RadioGroup'
