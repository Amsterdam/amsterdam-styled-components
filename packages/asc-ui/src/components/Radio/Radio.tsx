import React, { useState, useContext, useMemo } from 'react'
import RadioStyle, {
  RadioWrapperStyle,
  RadioCircleStyle,
  Props,
} from './RadioStyle'
import RadioContext from './RadioContext'
import LabelContext from '../Label/LabelContext'

const Radio: React.FC<Props & React.HTMLAttributes<HTMLDivElement>> = ({
  className,
  onChange,
  variant,
  disabled,
  id,
  defaultChecked,
  error: errorProp,
  ...otherProps
}) => {
  const [focus, setFocus] = useState(false)
  const {
    setChecked,
    checked: checkedProp,
    name,
    error: errorGroup,
  } = useContext(RadioContext)
  const { setActive } = useContext(LabelContext)
  const checked = checkedProp === id
  const error = errorProp || errorGroup || false

  // Make the label aware of changes in the checked state
  useMemo(() => {
    setActive(checked)
  }, [checked, setActive])

  // Set checked on defaultChecked
  if (defaultChecked && !checkedProp) {
    setChecked(id)
  }

  return (
    <RadioWrapperStyle
      {...{
        focus,
        error,
        className,
        variant,
        disabled,
        checked,
      }}
      aria-disabled={disabled}
    >
      <RadioCircleStyle
        {...{
          error,
          disabled,
          focus,
          variant,
          checked,
        }}
      />
      <RadioStyle
        {...{
          ...otherProps,
          disabled,
          id,
          defaultChecked,
          name,
        }}
        onFocus={() => {
          setFocus(true)
        }}
        onBlur={() => setFocus(false)}
        onChange={(e: React.ChangeEvent<HTMLInputElement>) => {
          if (onChange) {
            onChange(e)
          }
          setChecked(e.target.id)
        }}
      />
    </RadioWrapperStyle>
  )
}

export default Radio
