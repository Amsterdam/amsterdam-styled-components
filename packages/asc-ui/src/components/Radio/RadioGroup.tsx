import React from 'react'

import usePassPropsToChildren from '../../utils/hooks/usePassPropsToChildren'
import RadioGroupStyle, { Props } from './RadioGroupStyle'
import RadioContext from './RadioContext'

const RadioGroup: React.FC<Props & React.HTMLAttributes<HTMLDivElement>> = ({
  disabled,
  error,
  name,
  horizontal,
  children: childrenProps,
}) => {
  const [selected, setSelected] = React.useState('')

  const { children } = usePassPropsToChildren(childrenProps, {
    disabled,
  })
  return (
    <RadioContext.Provider
      value={{
        selected,
        setSelected,
        name,
        error,
      }}
    >
      <RadioGroupStyle horizontal={horizontal}>
        {disabled ? children : childrenProps}
      </RadioGroupStyle>
    </RadioContext.Provider>
  )
}

export default RadioGroup
