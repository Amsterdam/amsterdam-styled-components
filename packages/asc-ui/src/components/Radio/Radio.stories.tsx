import React, { createRef, useEffect } from 'react'
import Label from '../Label'
import Radio from './Radio'
import RadioGroup from './RadioGroup'
import Readme from './README.md'

export default {
  title: 'Experimental/Atoms/Radio',
  parameters: {
    notes: Readme,
  },
  decorators: [
    (storyFn: () => React.ReactNode) => (
      <div
        style={{
          display: 'flex',
          maxWidth: 400,
          flexDirection: 'column',
          alignItems: 'flex-start',
          padding: '40px 10px',
        }}
      >
        {storyFn()}
      </div>
    ),
  ],
}

export const Variants = () => (
  <>
    <RadioGroup name="group-1">
      <Label htmlFor="radio-1" label="Default">
        <Radio id="radio-1" checked />
      </Label>
      <Label htmlFor="radio-2" label="Default">
        <Radio id="radio-2" />
      </Label>
      <Label htmlFor="radio-3" label="Default">
        <Radio id="radio-3" />
      </Label>
    </RadioGroup>
    <br />
    <br />
    <RadioGroup name="group-2">
      <Label htmlFor="radio-4" label="Primary">
        <Radio id="radio-4" variant="primary" checked />
      </Label>
      <Label htmlFor="radio-5" label="Primary">
        <Radio id="radio-5" variant="primary" />
      </Label>
      <Label htmlFor="radio-6" label="Primary">
        <Radio id="radio-6" variant="primary" />
      </Label>
    </RadioGroup>
    <br />
    <br />
    <RadioGroup name="group-3">
      <Label htmlFor="radio-7" label="Secondary">
        <Radio id="radio-7" variant="secondary" checked />
      </Label>
      <Label htmlFor="radio-8" label="Secondary">
        <Radio id="radio-8" variant="secondary" />
      </Label>
      <Label htmlFor="radio-9" label="Secondary">
        <Radio id="radio-9" variant="secondary" />
      </Label>
    </RadioGroup>
  </>
)

export const Disabled = () => (
  <>
    <RadioGroup name="group-1" disabled>
      <Label htmlFor="radio-1" label="Group Disabled">
        <Radio id="radio-1" checked />
      </Label>
      <Label htmlFor="radio-2" label="Group Disabled">
        <Radio id="radio-2" />
      </Label>
      <Label htmlFor="radio-3" label="Group Disabled">
        <Radio id="radio-3" />
      </Label>
      <Label htmlFor="radio-4" label="Group Disabled">
        <Radio id="radio-4" />
      </Label>
    </RadioGroup>
    <br />
    <br />
    <RadioGroup name="group-2">
      <Label htmlFor="radio-5" label="Enabled">
        <Radio id="radio-5" />
      </Label>
      <Label htmlFor="radio-6" label="Enabled">
        <Radio id="radio-6" />
      </Label>
      <Label htmlFor="radio-7" label="Enabled">
        <Radio id="radio-7" />
      </Label>
      <Label htmlFor="radio-8" label="Single Disabled" disabled>
        <Radio id="radio-8" />
      </Label>
    </RadioGroup>
  </>
)

export const Error = () => (
  <>
    <RadioGroup name="group-1" error>
      <Label htmlFor="radio-1" label="Error">
        <Radio id="radio-1" />
      </Label>
      <Label htmlFor="radio-2" label="Error">
        <Radio id="radio-2" />
      </Label>
      <Label htmlFor="radio-3" label="Error">
        <Radio id="radio-3" />
      </Label>
      <Label htmlFor="radio-4" label="Error">
        <Radio id="radio-4" />
      </Label>
    </RadioGroup>
  </>
)

export const RadioGroupPositions = () => (
  <>
    <RadioGroup name="group-1" horizontal>
      <Label htmlFor="radio-1" label="Radio horizontal">
        <Radio id="radio-1" />
      </Label>
      <Label htmlFor="radio-2" label="Radio horizontal">
        <Radio id="radio-2" />
      </Label>
      <Label htmlFor="radio-3" label="Radio horizontal">
        <Radio id="radio-3" />
      </Label>
      <Label htmlFor="radio-4" label="Radio horizontal">
        <Radio id="radio-4" />
      </Label>
    </RadioGroup>
    <br />
    <br />
    <RadioGroup name="group-2">
      <Label htmlFor="radio-5" label="Radio vertical">
        <Radio id="radio-5" />
      </Label>
      <Label htmlFor="radio-6" label="Radio vertical">
        <Radio id="radio-6" />
      </Label>
      <Label htmlFor="radio-7" label="Radio vertical">
        <Radio id="radio-7" />
      </Label>
      <Label htmlFor="radio-8" label="Radio vertical">
        <Radio id="radio-8" />
      </Label>
    </RadioGroup>
  </>
)

export const RadioLabelPositions = () => (
  <>
    <RadioGroup name="group-1">
      <Label htmlFor="top" label="Top" position="top">
        <Radio id="top" />
      </Label>
      <Label htmlFor="right" label="Right" position="right">
        <Radio id="right" />
      </Label>
      <Label htmlFor="bottom" label="Bottom" position="bottom">
        <Radio id="bottom" />
      </Label>
      <Label htmlFor="left" label="Left" position="left">
        <Radio id="left" />
      </Label>
    </RadioGroup>
  </>
)

export const RadioWithRef = () => {
  const ref = createRef<HTMLInputElement>()

  useEffect(() => {
    const element = ref.current

    function onChange() {
      // eslint-disable-next-line no-console
      console.log('Value changed to option 2!')
    }

    if (!element) {
      return undefined
    }

    element.addEventListener('change', onChange)

    return () => {
      element.removeEventListener('change', onChange)
    }
  }, [ref])

  return (
    <RadioGroup name="group-1">
      <Label htmlFor="some-option-1" label="Option 1">
        <Radio id="some-option-1" checked />
      </Label>
      <Label htmlFor="some-option-2" label="Option 2">
        <Radio id="some-option-2" ref={ref} />
      </Label>
      <Label htmlFor="some-option-3" label="Option 3">
        <Radio id="some-option-3" />
      </Label>
    </RadioGroup>
  )
}
