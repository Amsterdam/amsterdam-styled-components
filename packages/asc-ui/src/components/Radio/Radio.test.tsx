import * as React from 'react'
import { render, cleanup, fireEvent } from '@testing-library/react'
import { ascDefaultTheme, ThemeProvider } from '@datapunt/asc-core'
import RadioGroup from './RadioGroup'
import Radio from './Radio'

describe('Radio', () => {
  let container: any

  const onChangeMock = jest.fn()

  beforeEach(() => {
    cleanup()
    ;({ container } = render(
      <ThemeProvider theme={ascDefaultTheme}>
        <RadioGroup id="group">
          <Radio onChange={onChangeMock} id="radio-1" />
          <Radio onChange={onChangeMock} id="radio-2" />
        </RadioGroup>
      </ThemeProvider>,
    ))
  })

  it('should render', () => {
    expect(container).toMatchSnapshot()
  })

  it('should match the style', () => {
    // Native radio should be hidden
    const radio = container.querySelector('#radio-1')
    expect(radio).toHaveStyleRule('opacity', '0')

    const circle = radio.previousSibling
    expect(circle).toHaveStyleRule('border-radius', '50%')

    const wrapper = radio.parentNode
    expect(wrapper).toHaveStyleRule('margin-bottom', '1px')

    const radioGroup = container.querySelector('#group')
    expect(radioGroup).toHaveStyleRule('flex-direction', 'column')
  })

  it('should change on click events', () => {
    const radio1 = container.querySelector('#radio-1')
    const circle1 = radio1.previousSibling
    const radio2 = container.querySelector('#radio-2')

    expect(circle1).toHaveStyleRule('color', '#767676')

    // Click on Radio 1
    fireEvent.click(radio1)

    expect(onChangeMock).toHaveBeenCalledTimes(1)
    expect(circle1).not.toHaveStyleRule('color', '#767676')

    // Clilck on Radio 2
    fireEvent.click(radio2)

    expect(onChangeMock).toHaveBeenCalledTimes(2)
    expect(circle1).toHaveStyleRule('color', '#767676')
  })
})
