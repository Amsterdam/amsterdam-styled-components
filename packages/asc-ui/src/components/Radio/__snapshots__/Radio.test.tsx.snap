// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Radio should render 1`] = `
.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
}

.c4 {
  top: 0;
  left: 0;
  width: 100%;
  cursor: inherit;
  height: 100%;
  margin: 0;
  padding: 0;
  position: absolute;
  opacity: 0;
}

.c3 {
  width: 24px;
  height: 24px;
  color: #767676;
  border: 1px solid;
  border-radius: 50%;
}

.c1 {
  position: relative;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  vertical-align: middle;
  -webkit-flex-shrink: 0;
  -ms-flex-negative: 0;
  flex-shrink: 0;
  padding: 4px;
  margin-bottom: 1px;
  margin-right: 8px;
}

.c1:hover .c2 {
  border: 2px solid;
  color: #000000;
}

<div>
  <div
    class="c0"
    id="group"
  >
    <div
      class="c1"
    >
      <span
        class="c2 c3"
      />
      <input
        class="c4"
        id="radio-1"
        type="radio"
      />
    </div>
    <div
      class="c1"
    >
      <span
        class="c2 c3"
      />
      <input
        class="c4"
        id="radio-2"
        type="radio"
      />
    </div>
  </div>
</div>
`;
